tests/cryptolibAESSHA/cryptolibAESSHA.ino

```diff
@@ -0,0 +1,137 @@
+  for (uint32_t j=0; j < data_size; j+=16){
+    for (uint8_t i = 0; i < 16; i++){
+      in[i] = data[j+i] ^ out[i];
+    }
+    cau_aes_encrypt (in, keysched, 10, out); // # 16-byte block
+    memcpy(&cipher_text[j],out,16);
+  }
+}
+
+void aes_cbc_decrypt(unsigned char *clear_text, const unsigned char *cipher_text){
+  //Data length should be a multiple of 16 bytes
+  // Need to initialize "iv" with the initialization vector
+  for (uint32_t j=0; j < data_size; j+=16){
+    cau_aes_decrypt (&cipher_text[j], keysched, 10, out); 
+    for (uint8_t i = 0; i < 16; i++){
+      clear_text[j+i] = out[i] ^ iv[i];
+    }
+    memcpy(iv,&cipher_text[j],16);
+  }
+}
+
+
+void loop() {
+  
+  Serial.println("SHA-256 State:");
+  cau_sha256_initialize_output(shastate);
+  for (i = 0; i < 8; i++) {
+    sprintf(str, "%08X ", shastate[i]);
+    Serial.print(str);
+  } Serial.println();
+  
+  t = micros();
+  cau_sha256_update (data, sizeof(data) / 64, shastate); // # 64-byte blocks
+  t = micros() - t;
+  
+  sprintf(str, "sha256: %d bytes in %u us   KBs  ", sizeof(data), t);
+  Serial.print(str);
+  Serial.println(1000.*sizeof(data) / t);
+  for (i = 0; i < 8; i++) {
+    sprintf(str, "%08x ", shastate[i]);
+    Serial.print(str);
+  } Serial.println();
+
+  Serial.println("AES-128:");
+  t = micros();
+  cau_aes_set_key(aeskey, 128, keysched);
+  t = micros() - t;
+  Serial.print("aes set key microsec "); Serial.println(t);
+  //printf("aes set key  %u us\n",t);
+  t = micros();
+  cau_aes_encrypt (in, keysched, 10, cipher_text); // # 16-byte block
+  t = micros() - t;
+  sprintf(str, "aes %d bytes %u us  KBs  ", sizeof(in), t); Serial.print(str);
+  Serial.println(1000.*sizeof(in) / t);
+  cau_aes_decrypt (cipher_text, keysched, 10, iv); //  decrypt test
+  errs = 0;
+  for (i = 0; i < 16; i++) if (in[i] != iv[i]) errs++;
+  Serial.print("aes errs "); Serial.println(errs);
+  
+  // CBC XOR init_vector or cipher_textput with plain, our sketch does 4 blocks
+  Serial.print("\nAES Key: ");
+  for (i = 0; i < sizeof(aeskey); i++) {
+    sprintf(str, "%02x", aeskey[i]);
+    Serial.print(str);
+  } Serial.println();
+  
+  memcpy(out,init_vector,16);
+  Serial.print("AES IV: ");
+  for (i = 0; i < sizeof(init_vector); i++) {
+    sprintf(str, "%02x", init_vector[i]);
+    Serial.print(str);
+  } Serial.println();
+
+  Serial.print("Plain  Text: ");
+  for (i = 0; i < sizeof(data); i++) {
+    sprintf(str, "%02x", data[i]);
+    Serial.print(str);
+  } Serial.println();
+  
+  t = micros();
+  aes_cbc_encrypt(data,cipher_text);
+  t = micros() - t;
+  Serial.print("Cipher Text: ");
+  for (i = 0; i < sizeof(cipher_text); i++) {
+    sprintf(str, "%02x", cipher_text[i]);
+    Serial.print(str);
+  } Serial.println();
+
+   
+  sprintf(str, "aes cbc encrypt %d bytes %u us, KBs ", sizeof(data), t);
+  Serial.print(str);
+  Serial.println(1000.*sizeof(data) / t);
+  
+  t = micros();
+  memcpy(iv,init_vector,16);
+  aes_cbc_decrypt(clear_text, cipher_text);
```

Comment:
you're testing 'round-trip' through your aes-cbc routines; this is a good test.

It is necessary but not sufficient. You should also test with known test vectors. e.g. https://github.com/google/wycheproof/tree/master/testvectors

Reply:
Looks interesting. I'll have to look into this more. It looks challenging to port...

Reply:
No need to port the project; but you should include tests that confirm that your implementation here is capabale of encrypting a known plaintext into and expected ciphertext and the analogous test for decrypt.  Right now your tests are only confirming that you have a pair of functions that are a matching encrypt/decrypt; but you're not confirming that existing tools/implementations could encrypt or decrypt the outputs of your implementations

Reply:
We tested with the online calculator at https://www.cryptogrium.com/aes-cbc.html and obtained matching results.

Reply:
You should use the encrypted output of that online calculator as the input in a test here (and vice versa). Right now your tests verify that your implementation can do a round trip. You are missing a test that to confirm that the implementation will 'work with standard tools'.